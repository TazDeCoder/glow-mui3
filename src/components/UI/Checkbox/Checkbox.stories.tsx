import type { ComponentMeta, ComponentStory } from '@storybook/react';

import { Checkbox } from './Checkbox';
import { Box } from '../../Utils';

export default {
  title: 'Components/UI/Checkbox',
  component: Checkbox,
  argTypes: {
    onClick: { action: 'toggled' },
  },
} as ComponentMeta<typeof Checkbox>;

export const Default: ComponentStory<typeof Checkbox> = (args) => (
  <Checkbox label="Label" {...args} />
);

export const Colors: ComponentStory<typeof Checkbox> = (args) => (
  <Box
    sx={{
      display: 'flex',
      flexDirection: 'column',
      rowGap: 2,
    }}
  >
    <Checkbox {...args} color="primary" label="Primary" />
    <Checkbox {...args} color="secondary" label="Secondary" />
    <Checkbox {...args} color="info" label="Info" />
    <Checkbox {...args} color="error" label="Error" />
    <Checkbox {...args} color="warning" label="Warning" />
    <Checkbox {...args} color="success" label="Success" />
  </Box>
);
