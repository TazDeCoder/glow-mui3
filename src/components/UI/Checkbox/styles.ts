import type { ThemeUIStyleObject } from '@theme-ui/core';

export const labelStyles: ThemeUIStyleObject = {
  display: 'flex',
  alignItems: 'center',
  gap: 12,
};

export const baseStyles: ThemeUIStyleObject = {
  position: 'relative',
  display: 'inline-block',
  width: 18,
  height: 18,
  border: '2px solid',
  borderRadius: 2,
  transition: 'standard',
};

export const checkedStyles: ThemeUIStyleObject = {
  '&:hover::after': {
    content: '""',
    position: 'absolute',
    top: '-2px',
    left: '-2px',
    width: 'calc(100% + 4px)',
    height: 'calc(100% + 4px)',
    bg: 'action.hover',
    borderRadius: 'inherit',
  },
};

export const iconStyles: ThemeUIStyleObject = {
  fill: 'none',
  strokeWidth: 3,
};
