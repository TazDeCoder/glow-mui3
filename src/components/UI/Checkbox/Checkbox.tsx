/** @jsxRuntime classic */
/** @jsx jsx */

import { forwardRef, useState } from 'react';
import { jsx, ThemeUIStyleObject } from '@theme-ui/core';
import { hideVisually } from 'polished';

import { Box, BoxOwnProps } from '../../Utils';
import { Typography } from '../Typography';

import { baseStyles, checkedStyles, labelStyles, iconStyles } from './styles';
import type { Assign, ForwardRef, ColorTheme } from '../../types';

export interface CheckboxProps
  extends Assign<React.ComponentPropsWithRef<'input'>, BoxOwnProps> {
  /** Color theme used for the component
   * @default 'primary'
   */
  color?: ColorTheme;
  /** Label associated to the component */
  label?: string;
}

export const Checkbox: ForwardRef<HTMLInputElement, CheckboxProps> = forwardRef(
  ({ color = 'primary', label, defaultChecked = false, sx, ...props }, ref) => {
    const [checked, setChecked] = useState(defaultChecked);

    const handleChange = (evt: React.ChangeEvent<HTMLInputElement>) => {
      setChecked(evt.target.checked);
    };

    const stateStyles: ThemeUIStyleObject = {
      ...(checked && {
        ...checkedStyles,
        bg: `${color}.main`,
        borderColor: `${color}.main`,
      }),
      ...(!checked && {
        borderColor:
          color === 'primary' || color === 'secondary'
            ? 'surface.onMain'
            : `${color}.main`,
      }),
    };

    return (
      <Box sx={{ width: 'min-content' }}>
        <Box
          as="label"
          sx={{
            ...labelStyles,
            ...sx,
          }}
        >
          <Box
            sx={{
              display: 'inline-block',
              cursor: 'pointer',
              verticalAlign: 'top',
            }}
          >
            <Box
              ref={ref}
              as="input"
              type="checkbox"
              checked={checked}
              onChange={handleChange}
              sx={{ ...hideVisually() }}
              {...props}
            />
            <Box
              sx={{
                ...baseStyles,
                ...stateStyles,
              }}
            >
              <Box
                as="svg"
                sx={{
                  ...iconStyles,
                  stroke: checked && `${color}.onMain`,
                  visibility: checked ? 'visible' : 'hidden',
                }}
                {...({ viewBox: '0 0 24 24' } as {})}
              >
                <polyline points="20 6 9 17 4 12" />
              </Box>
            </Box>
          </Box>
          {label && (
            <Typography variant="label" size="lg" color="bg.onMain">
              {label}
            </Typography>
          )}
        </Box>
      </Box>
    );
  },
);
