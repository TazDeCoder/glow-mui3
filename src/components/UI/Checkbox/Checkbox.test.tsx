import { act, render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { Checkbox } from './Checkbox';

describe('<Checkbox />', () => {
  describe('Snapshots', () => {
    it('should render correctly', () => {
      const { container } = render(<Checkbox />);
      expect(container).toMatchSnapshot();
    });
  });
  describe('User events', () => {
    it('triggers checked state to true when toggled once', async () => {
      const { getByRole } = render(<Checkbox />);
      const checkboxEl = getByRole('checkbox');
      await act(async () => {
        await userEvent.click(checkboxEl);
      });
      expect(checkboxEl).toBeChecked();
    });
    it('triggers checked state back to false when toggled twice', async () => {
      const { getByRole } = render(<Checkbox />);
      const checkboxEl = getByRole('checkbox');
      await act(async () => {
        await userEvent.click(checkboxEl);
        await userEvent.click(checkboxEl);
      });
      expect(checkboxEl).not.toBeChecked();
    });
  });
});
