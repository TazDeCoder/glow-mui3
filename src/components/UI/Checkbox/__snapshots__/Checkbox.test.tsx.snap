// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Checkbox /> Snapshots should render correctly 1`] = `
.emotion-0 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  width: -webkit-min-content;
  width: -moz-min-content;
  width: min-content;
}

.emotion-1 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 12px;
}

.emotion-2 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  display: inline-block;
  cursor: pointer;
  vertical-align: top;
}

.emotion-3 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  border: 0;
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

.emotion-4 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  position: relative;
  display: inline-block;
  width: 18px;
  height: 18px;
  border: 2px solid;
  border-radius: 2px;
  -webkit-transition: standard;
  transition: standard;
  border-color: surface.onMain;
}

.emotion-5 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  fill: none;
  stroke-width: 3;
  visibility: hidden;
}

<div>
  <div
    class="emotion-0"
  >
    <label
      class="emotion-1"
    >
      <div
        class="emotion-2"
      >
        <input
          class="emotion-3"
          type="checkbox"
        />
        <div
          class="emotion-4"
        >
          <svg
            class="emotion-5"
            viewBox="0 0 24 24"
          >
            <polyline
              points="20 6 9 17 4 12"
            />
          </svg>
        </div>
      </div>
    </label>
  </div>
</div>
`;
